lib/HeroBonus.cpp

```diff
@@ -1721,24 +1724,49 @@ bool CPropagatorNodeType::shouldBeAttached(CBonusSystemNode *dest)
 	return nodeType == dest->getNodeType();
 }
 
-CreatureNativeTerrainLimiter::CreatureNativeTerrainLimiter(int TerrainType)
+CreatureTerrainLimiter::CreatureTerrainLimiter(int TerrainType)
 	: terrainType(TerrainType)
 {
 }
 
-CreatureNativeTerrainLimiter::CreatureNativeTerrainLimiter()
+CreatureTerrainLimiter::CreatureTerrainLimiter()
 	: terrainType(-1)
 {
 
 }
 
-int CreatureNativeTerrainLimiter::limit(const BonusLimitationContext &context) const
+int CreatureTerrainLimiter::limit(const BonusLimitationContext &context) const
 {
-	const CCreature *c = retrieveCreature(&context.node);
-	return !c || !c->isItNativeTerrain(terrainType); //drop bonus for non-creatures or non-native residents
+	const CStack *stack = retrieveStackBattle(&context.node);
+	if(stack && stack->base)
+	{
+		if(terrainType == -1)//terrainType not specified = native
+			return !stack->isOnNativeTerrain();
+		return stack->base->armyObj->battle->terrainType != terrainType;
+	}
+	return true;
 	//TODO neutral creatues
 }
 
+std::string CreatureTerrainLimiter::toString() const
+{
+	char buf[100];
+	sprintf(buf, "CreatureTerrainLimiter(terrainType=%s)",

```

Comment:
If you really want to use `printf` instead of boost::format, use more safer `snprintf` and make 100 a named constant